lib/honeycomb/client.rb

```diff
@@ -44,11 +47,35 @@ def initialize(configuration:)
       end
     end
 
-    def start_span(name:, serialized_trace: nil, **fields)
+    def propagation_context_from_req(env:)
+      custom_hook = @additional_trace_options[:parser_hook]
+      if custom_hook.nil?
+        parser = Honeycomb::HoneycombPropagation::Parser.new
+        parser.unmarshal_trace_context(env)
+      else
+        custom_hook.call(env)
+      end
+    end
+
+    def header_from_propagation_context(propagation_context)
+      custom_hook = @additional_trace_options[:propagation_hook]
+      if custom_hook.nil?
+        propagator = Honeycomb::HoneycombPropagation::Propagator.new
+        propagator.marshal_trace_context(propagation_context)
+      else
+        custom_hook.call(propagation_context)
+      end
+    end
+
+    def start_span(
+      name:, serialized_trace: nil, propagation_context: nil,
```

Comment:
Feels a bit weird that we're adding an optional `propagation_context` argument to every `start_span` call, instead of having a specific `propagate_and_start_trace` call for when propagation is happening.

For example, what should happen if you pass in a propagation context and there's already a trace? It feels like there are two explicit use cases, and this seems to be mixing them.

Reply:
Yeah just thinking about this one. This is another one of those places where an existing flow and I mostly mimicked it, but I'm not sure the old flow is necessarily a problem? (I'm really not sure about this though.) The `Client.start_span` is kind of a misleading-ish name (probably could be `propagate_and_start_trace`, because the actual Span is initialized in `Trace` (which is initialized here). `Span` is of course the most central instance here, and the trace_id, parent_id, etc need to be available to all these instances without spinning up a new instance.

On that specific example, there's a check for `context.current_trace.nil?` before initializing a Trace (that receives a propagation_context), otherwise the `create_child` method is called on the current instance of `Span`.